SYCL/ESIMD/lsc/lsc_argument_type_deduction.cpp

```diff
@@ -0,0 +1,114 @@
+//==-------- lsc_argument_type_deduction.cpp - DPC++ ESIMD on-device test --==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu-intel-pvc || esimd_emulator
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+
+// The test checks that compiler is able to corrctly deduce the template
+// parameter types for lsc functions.
+
+#include "../esimd_test_utils.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::intel::esimd;
+using namespace sycl::ext::intel::experimental::esimd;
+
+template <unsigned SIMDSize> int testAccessor(queue q) {
+  auto size = size_t{128};
+
+  auto vec_0 = std::vector<int>(size);
+
+  std::iota(vec_0.begin(), vec_0.end(), 0);
+  auto buf_0 = buffer{vec_0};
+
+  try {
+    q.submit([&](handler &h) {
+      auto access_0 = buf_0.template get_access<access::mode::read_write>(h);
+
+      h.parallel_for(
+          range<1>{size / SIMDSize}, [=](id<1> id) SYCL_ESIMD_KERNEL {
+            auto offset = id[0] * SIMDSize * sizeof(int);
+            auto offsets = simd<uint32_t, SIMDSize>(id * SIMDSize * sizeof(int),
+                                                    sizeof(int));
+            simd<int, SIMDSize> data_0 = lsc_block_load(access_0, offset);
```

Comment:
This is very suspicious place. `NElts` template argument inside `lsc_block_load`  implementation can't be deduced from the data_0 type, AFAIU. E.g. this does not compile with clang c++17:

```cpp
template <class T, int N> struct S{};

template <class T, int N> S<T, N> foo(T* ptr) {
    return S<T, N>();
}

void bar(float *ptr) {
    S<float, 10> x = foo(ptr);
}
```

please double check what happens here